src/herder/SurgePricingUtils.cpp

```diff
@@ -41,10 +50,550 @@ feeRate3WayCompare(int64 lFeeBid, uint32 lNbOps, int64 rFeeBid, uint32 rNbOps)
     return 0;
 }
 
-int
-feeRate3WayCompare(TransactionFrameBase const& l, TransactionFrameBase const& r)
+int64_t
+computeBetterFee(TransactionFrameBase const& tx, int64_t refFeeBid,
+                 uint32_t refNbOps)
+{
+    constexpr auto m = std::numeric_limits<int64_t>::max();
+
+    int64 minFee = m;
+    int64 v;
+    if (bigDivide(v, refFeeBid, tx.getNumOperations(), refNbOps,
+                  Rounding::ROUND_DOWN) &&
+        v < m)
+    {
+        minFee = v + 1;
+    }
+    return minFee;
+}
+
+SurgePricingPriorityQueue::TxStackComparator::TxStackComparator(bool isGreater,
+                                                                size_t seed)
+    : mIsGreater(isGreater), mSeed(seed)
+{
+}
+
+bool
+SurgePricingPriorityQueue::TxStackComparator::operator()(
+    TxStackPtr const& txStack1, TxStackPtr const& txStack2) const
+{
+    return txStackLessThan(*txStack1, *txStack2) ^ mIsGreater;
+}
+
+bool
+SurgePricingPriorityQueue::TxStackComparator::compareFeeOnly(
+    TransactionFrameBase const& tx1, TransactionFrameBase const& tx2) const
+{
+    return compareFeeOnly(tx1.getFeeBid(), tx1.getNumOperations(),
+                          tx2.getFeeBid(), tx2.getNumOperations());
+}
+
+bool
+SurgePricingPriorityQueue::TxStackComparator::compareFeeOnly(
+    int64_t tx1Bid, uint32_t tx1Ops, int64_t tx2Bid, uint32_t tx2Ops) const
+{
+    bool isLess = feeRate3WayCompare(tx1Bid, tx1Ops, tx2Bid, tx2Ops) < 0;
+    return isLess ^ mIsGreater;
+}
+
+bool
+SurgePricingPriorityQueue::TxStackComparator::isGreater() const
+{
+    return mIsGreater;
+}
+
+bool
+SurgePricingPriorityQueue::TxStackComparator::txStackLessThan(
+    TxStack const& txStack1, TxStack const& txStack2) const
+{
+    if (txStack1.empty())
+    {
+        return !txStack2.empty();
+    }
+    if (txStack2.empty())
+    {
+        return false;
+    }
+    return txLessThan(txStack1.getTopTx(), txStack2.getTopTx(), true);
+}
+
+bool
+SurgePricingPriorityQueue::TxStackComparator::txLessThan(
+    TransactionFrameBaseConstPtr const& tx1,
+    TransactionFrameBaseConstPtr const& tx2, bool breakTiesWithHash) const
+{
+    auto cmp3 = feeRate3WayCompare(*tx1, *tx2);
+
+    if (cmp3 != 0 || !breakTiesWithHash)
+    {
+        return cmp3 < 0;
+    }
+    // break tie with pointer arithmetic
+    auto lx = reinterpret_cast<size_t>(tx1.get()) ^ mSeed;
+    auto rx = reinterpret_cast<size_t>(tx2.get()) ^ mSeed;
+    return lx < rx;
+}
+
+SurgePricingPriorityQueue::SurgePricingPriorityQueue(
+    bool isHighestPriority, std::shared_ptr<SurgePricingLaneConfig> settings,
+    size_t comparisonSeed)
```

Comment:
This is definitely not the order I have arrived at these changes as some refactoring choices are not obvious without considering the lanes. My line of thinking was to do the change that is non-functional if the flag is disabled (which is hopefully verified by the tests). However, I realize this may be too hard to review. Let me try to separate this into a couple PRs. I would propose:
- dex ops sepration
- Refactoring to use `SurgePricingPriorityQueue` everywhere, with only a 'generic' lane
- add dex lane/config

There is no 'code move' step as this is not what is happening - I'm removing 3x duplication of very similar code and replacing it with a new data structure. The refactoring step will probably remain pretty chunky, but that's because the refactoring itself is complicated. I really prefer to trust the tests to ensure the changes are functionally a no-op - otherwise no significant refactoring would be possible.

Please let me know if that sounds reasonable.

Reply:
If you're going to split this up into multiple PRs, I would actually do it as:
* PR 1  (probably repurpose this one?) -> Refactoring to use SurgePricingPriorityQueue everywhere, with only a 'generic' lane.
    * if you can't have simpler "code move" commits, you should separate commits into logical chunks. I know in the existing PR it's pretty hard to understand which part of the code is being effected because "everything changes" (flooding, nomination and limits at the same time)
* PR 2 -> dex ops separation and add dex lane/config (I don't think you can split this into 2 PRs as "dex ops separation" is not used outside of this change)